klient/k8s/resources/decoder.go

```diff
@@ -0,0 +1,293 @@
+// // to each object after they have been decoded.
+// func DecodeObjectsByGVK(manifest io.Reader, kind schema.GroupVersionKind, patches ...MutateFunc) ([]k8s.Object, error) {
+// 	objects := []k8s.Object{}
+// 	// start decoding documents
+// 	decoder := yaml.NewYAMLOrJSONDecoder(manifest, 1024)
+// 	for {
+// 		o, err := scheme.Scheme.New(kind)
+// 		if err != nil {
+// 			return objects, err
+// 		}
+// 		obj := k8s.Object(&object{metav1.ObjectMeta{}, o})
+// 		if err := decoder.Decode(obj); errors.Is(err, io.EOF) {
+// 			break
+// 		} else if err != nil {
+// 			return nil, err
+// 		}
+// 		for _, patch := range patches {
+// 			patch(obj)
+// 		}
+// 		objects = append(objects, obj)
+// 	}
+// 	return objects, nil
+// }
+
+// DecodeListItems decodes a multi-document YAML or JSON file into items on the provided k8s.ObjectList. Patches are applied
+// to each object after they have been decoded.
+func DecodeListItems(manifest io.Reader, obj k8s.ObjectList, patches ...MutateFunc) error {
+	// get a pointer to the list's Items slice
+	itemsPtr, err := meta.GetItemsPtr(obj)
+	if err != nil {
+		return err
+	}
+	// convert to pointer so we can append items
+	items, err := conversion.EnforcePtr(itemsPtr)
+	if err != nil {
+		return err
+	}
+	// determine the type of the slice Items and create a new reference object as a base to decode into for each document
+	base := reflect.New(items.Type().Elem()).Interface().(k8s.Object)
+	// start decoding documents
+	decoder := yaml.NewYAMLOrJSONDecoder(manifest, 1024)
+	for {
+		obj := base.DeepCopyObject().(k8s.Object) // copy the base object to decode new object to
+		if err := decoder.Decode(obj); errors.Is(err, io.EOF) {
+			break
+		} else if err != nil {
+			return err
+		}
+		for _, patch := range patches {
+			patch(obj)
+		}
+		items.Set(reflect.Append(items, reflect.ValueOf(obj).Elem()))
+	}
+	return nil
+}
+
+// DecodeDocuments decodes a multi-document YAML or JSON file into a copy of the provided k8s.Object and invokes
+// fn on each decoded object, after patches have been applied.
+func DecodeDocuments(manifest io.Reader, base k8s.Object, fn func(k8s.Object), patches ...MutateFunc) error {
+	// start decoding documents
+	decoder := yaml.NewYAMLOrJSONDecoder(manifest, 1024)
+	for {
+		obj := base.DeepCopyObject().(k8s.Object) // copy the base object to decode new object to
+		if err := decoder.Decode(obj); errors.Is(err, io.EOF) {
+			break
+		} else if err != nil {
+			return err
+		}
+		for _, patch := range patches {
+			patch(obj)
+		}
+		fn(obj)
+	}
+	return nil
+}
+
+// Decoder decodes a multi-document YAML or JSON file into Go objects for any types that have been registered with scheme.Scheme.
+// If the GroupVersionKind defaults is provided, it is used when determining the object Kind.
+func Decoder(manifest io.Reader, defaults *schema.GroupVersionKind, fn func(k8s.Object), patches ...MutateFunc) error {
```

Comment:
Maybe rename `Decoder` to `DecodeAll`.
Also, does it decode a document with multi-objects document or multi-documents/multi-objects ?

Reply:
Also, what's the difference (in argument list) between `fn` and `patches` they seem to do the same thing, mainly ability to update the object post parsing.

Reply:
> Also, does it decode a document with multi-objects document or multi-documents/multi-objects ?

These functions above Decode (Decoder, DecodeDocuments, DecodeListItems,  DecodeObjects, and DecodeDirectory) all support multi-document YAML files. I am not sure I am parsing this question correctly. (YAML stream might be the right term https://yaml.org/spec/1.2.2/#92-streams)

(additional context: these are all solving the same use-case, minus DecodeDirectory, but with a different function signature)

> Also, what's the difference (in argument list) between `fn` and `patches` they seem to do the same thing, mainly ability to update the object post parsing.

Fair point -- they are fairly close but:
- patches are optional and apply changes to an object before it is called by the `fn` func. Patches seemed useful from the single-document examples, so I carried them over.
- Inspiration: `meta.EachListItem(obj runtime.Object, fn func(runtime.Object) error) error` [meta.EachListItem](https://pkg.go.dev/k8s.io/apimachinery/pkg/api/meta#EachListItem)

`fn` is meant to actually _use_ the object; this function doesn't return anything, but allows you to pass in a handler that can execute based on the object.